src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -183,230 +139,165 @@ uct_cuda_ipc_get_remote_cache(pid_t pid, uct_cuda_ipc_cache_t **cache)
-        status = UCS_PROFILE_CALL(ucs_pgtable_insert,
-                                  &cache->pgtable, &region->super);
-    }
-    if (status != UCS_OK) {
+    return status;
+}
 
-        ucs_error("%s: failed to insert region:"UCS_PGT_REGION_FMT" size:%lu :%s",
-                  cache->name, UCS_PGT_REGION_ARG(&region->super), key->b_len,
-                  ucs_status_string(status));
-        ucs_free(region);
-        goto err;
-    }
+static ucs_status_t
+uct_cuda_ipc_rcache_mem_reg(void *context, ucs_rcache_t *rcache, void *arg,
+                            ucs_rcache_region_t *region, uint16_t flags)
+{
+    uct_cuda_ipc_key_t *key                       = (uct_cuda_ipc_key_t*)arg;
+    uct_cuda_ipc_rcache_region_t *cuda_ipc_region =
+                    ucs_derived_of(region, uct_cuda_ipc_rcache_region_t);
 
-    ucs_trace("%s: cuda_ipc cache new region:"UCS_PGT_REGION_FMT" size:%lu",
-              cache->name, UCS_PGT_REGION_ARG(&region->super), key->b_len);
+    cuda_ipc_region->ipc_handle = key->ph;
 
-    status = UCS_OK;
+    return uct_cuda_ipc_open_memhandle(&cuda_ipc_region->mapping_start, key);
 
-err:
-    pthread_rwlock_unlock(&cache->lock);
-    return status;
 }
 
-ucs_status_t uct_cuda_ipc_create_cache(uct_cuda_ipc_cache_t **cache,
-                                       const char *name)
+static void uct_cuda_ipc_rcache_mem_dereg(void *context, ucs_rcache_t *rcache,
+                                          ucs_rcache_region_t *region)
+{
+    uct_cuda_ipc_rcache_region_t *cuda_ipc_region =
+                    ucs_derived_of(region, uct_cuda_ipc_rcache_region_t);
+
+    uct_cuda_ipc_close_memhandle(cuda_ipc_region->mapping_start);
+
+}
+
+static void uct_cuda_ipc_rcache_dump_region(void *context, ucs_rcache_t *rcache,
+                                            ucs_rcache_region_t *region, char *buf,
+                                            size_t max)
+{
+    uct_cuda_ipc_rcache_region_t *cuda_ipc_region =
+                    ucs_derived_of(region, uct_cuda_ipc_rcache_region_t);
+
+    snprintf(buf, max, "mapping_start %p", cuda_ipc_region->mapping_start);
+}
+
+static ucs_rcache_ops_t uct_cuda_ipc_rcache_ops = {
+    .mem_reg     = uct_cuda_ipc_rcache_mem_reg,
+    .mem_dereg   = uct_cuda_ipc_rcache_mem_dereg,
+    .dump_region = uct_cuda_ipc_rcache_dump_region
+};
+
+
+ucs_status_t
+uct_cuda_ipc_create_cache(uct_cuda_ipc_md_t *md, ucs_rcache_t **cache,
+                          const char *name)
 {
     ucs_status_t status;
-    uct_cuda_ipc_cache_t *cache_desc;
-    int ret;
+    ucs_rcache_t *cache_desc;
+    ucs_rcache_params_t rcache_params;
 
-    cache_desc = ucs_malloc(sizeof(uct_cuda_ipc_cache_t), "uct_cuda_ipc_cache_t");
+    cache_desc = ucs_malloc(sizeof(struct ucs_rcache), "ucs_rcache");
     if (cache_desc == NULL) {
-        ucs_error("failed to allocate memory for cuda_ipc cache");
+        ucs_error("failed to allocate memory for cuda_ipc rcache");
         return UCS_ERR_NO_MEMORY;
     }
```

Comment:
Please remove these lines and `cache_desc` variable. `ucs_rcache_create` allocates memory for rcache.